news/321.feature

```diff
@@ -0,0 +1,8 @@
+- Add ability to nest interpolations, e.g. ${env:{$var}} or ${foo.${bar}.${baz}}
+- The `env` resolver does not cache its output anymore, i.e. `${env:VAR}` will yield a new value if environment variable `VAR` is modified between consecutive accesses
+- The `env` resolver now attempts to parse the string representation of environment variables in a way that is consistent with the interpolation grammar. This enables more advanced use cases (e.g. setting environment variables that represent lists or dictionaries, possibly with interpolations: `export VAR='[1, 2, {foo: ${foo}}]'`).
+- `register_resolver()` has new arguments to enable more advanced custom resolver behavior:
+    * `args_as_strings=False` makes it possible for custom resolvers to take non-string inputs
+    * `use_cache=False` disables the cache
+    * `config_arg` can be used to give the resolver access to the whole config object
+    * `parent_arg` can be used to give the resolver access to the parent of the key being processed
```

Comment:
I'm pretty sure this is not a proper news item, but all examples in the repo were only one-liners so I wasn't sure how to list multiple new things. Please let me know what is the correct way :)

Reply:
news/321.feature.1
news/321.feature.2
news/321.feature.3
each one of those will have a link to the actual issue (#321) so keep it short.

> The `env` resolver does not cache its output anymore
Did we agree to change this behavior? I will review the whole thing when I get to it, but I thought we didn't say this changes.

Also, if you are breaking compatibility (such as in this case, the file suffix for that item should be `removal`.

Reply:
> news/321.feature.1
> news/321.feature.2
> news/321.feature.3

Thanks! Will update the PR accordingly.

> Did we agree to change this behavior? I will review the whole thing when I get to it, but I thought we didn't say this changes.

Not explicitly. It's a consequence of making `env` parse strings with the grammar => it can now resolve interpolations if your environment variable contains some => its output can now change as soon as the config change, even with the same input and the same value of the environment variable => I disabled the cache as per our discussion in #266

Reply:
My [comment](https://github.com/omry/omegaconf/issues/266#issuecomment-673654968) is saying "I think having a use_cache flag is solving the problem".

The cache is critical for some existing use cases, we need to keep it.
We discussed adding a flag to disable the cache, not to kill it outright.

Reply:
> We discussed adding a flag to disable the cache, not to kill it outright.

Oh but I didn't kill it. I added a flag `use_cache`, which is set to False for resolvers who ask to access the config object. This is the case of the `env` resolver now, because it needs it to be able to parse the environment variable. This also seemed better to me as otherwise the two following scenarios would behave differently:

* Scenario 1: `bar` changes when `foo` changes:
```yaml
foo: 0
bar: ${foo}
```

* Scenario 2: `bar` would not change when `foo` changes if we used the cache for `env`:
```yaml
foo: 0
bar: ${env:BAR}
```
`export BAR=\${foo}` (escape is for the shell)